### PR TITLE
Add ONNX model support

### PR DESCRIPTION
**Pull request by thigm85**
I confirm that this contribution is made under the terms of the license found in the root directory of this repository's source tree and that I have the authority necessary to make this contribution on behalf of its copyright owner.

### CODE REVIEW

File: vespa/test_integration_docker.py [hunk 1]
```diff
@@ -141,3 +150,125 @@ def tearDown(self) -> None:
...
+                        "        0,\n"
+                        "    if (d1 < question_length + doc_length,\n"
+                        "        1,\n"
+                        "        TOKEN_NONE\n"
+                        "    )))",
+                    ),
+                    Function(
+                        name="eval",
+                        expression="tensor(x{}):{x1:onnxModel(bert_tiny).logits{d0:0,d1:0}}",
+                    ),
```
Comment (lesters) [inline vespa/test_integration_docker.py:245]: Can be simplified now to `onnx(bert_tiny).logits{d0:0,d1:0}`. `onnx` and `onnxModel` are now synonymous. Actually, you don't need this `eval` funciton at all, can just put it directly in second phase.

File: vespa/test_integration_vespa_cloud.py [hunk 2]
```diff
@@ -130,7 +136,135 @@ def test_data_operation(self):
...
+                        "        0,\n"
+                        "    if (d1 < question_length + doc_length,\n"
+                        "        1,\n"
+                        "        TOKEN_NONE\n"
+                        "    )))",
+                    ),
+                    Function(
+                        name="eval",
+                        expression="tensor(x{}):{x1:onnxModel(bert_tiny).logits{d0:0,d1:0}}",
+                    ),
```
Comment (lesters) [inline vespa/test_integration_vespa_cloud.py:237]: See comment above on docker test

File: vespa/test_package.py [hunk 3]
```diff
@@ -398,6 +429,77 @@ def setUp(self) -> None:
...
+                            "    if (d1 < question_length,\n"
+                            "        0,\n"
+                            "    if (d1 < question_length + doc_length,\n"
+                            "        1,\n"
+                            "        TOKEN_NONE\n"
+                            "    )))",
+                        ),
+                        Function(
+                            name="eval",
+                            expression="tensor(x{}):{x1:onnxModel(bert).logits{d0:0,d1:0}}",
```
Comment (lesters) [inline vespa/test_package.py:481]: same as above

File: vespa/test_package.py [hunk 4]
```diff
@@ -455,6 +564,74 @@ def test_schema_to_text(self):
...
+            "                        0,\n"
+            "                    if (d1 < question_length + doc_length,\n"
+            "                        1,\n"
+            "                        TOKEN_NONE\n"
+            "                    )))\n"
+            "            }\n"
+            "        }\n"
+            "        function eval() {\n"
+            "            expression {\n"
+            "                tensor(x{}):{x1:onnxModel(bert).logits{d0:0,d1:0}}\n"
```
Comment (lesters) [inline vespa/test_package.py:618]: same as above

File: vespa/test_integration_docker.py [hunk 5]
```diff
@@ -141,3 +150,125 @@ def tearDown(self) -> None:
...
+                functions=[
+                    Function(
+                        name="question_length",
+                        expression="sum(map(query(query_token_ids), f(a)(a > 0)))",
+                    ),
+                    Function(
+                        name="doc_length",
+                        expression="sum(map(attribute(doc_token_ids), f(a)(a > 0)))",
+                    ),
+                    Function(
```
Comment (lesters) [inline vespa/test_integration_docker.py:212]: Vespa now has a `tokenInputIds` function that does this. Can be replaced with that.

File: vespa/test_integration_docker.py [hunk 6]
```diff
@@ -141,3 +150,125 @@ def tearDown(self) -> None:
...
+                        "        TOKEN_SEP,\n"
+                        "        TOKEN_NONE\n"
+                        "    ))))))",
+                    ),
+                    Function(
+                        name="attention_mask",
+                        expression="map(input_ids, f(a)(a > 0)) ",
+                    ),
+                    Function(
+                        name="token_type_ids",
```
Comment (lesters) [inline vespa/test_integration_docker.py:233]: Similarly, Vespa has a `tokenTypeIds` function now.

File: vespa/test_integration_docker.py [hunk 7]
```diff
@@ -141,3 +150,125 @@ def tearDown(self) -> None:
...
+                        "        TOKEN_SEP,\n"
+                        "    if (d1 < question_length + doc_length + 2,\n"
+                        "        attribute(doc_token_ids){d0:(d1-question_length-2)},\n"
+                        "    if (d1 == question_length + doc_length + 2,\n"
+                        "        TOKEN_SEP,\n"
+                        "        TOKEN_NONE\n"
+                        "    ))))))",
+                    ),
+                    Function(
+                        name="attention_mask",
```
Comment (lesters) [inline vespa/test_integration_docker.py:229]: Similarly, Vespa has a `tokenAttentionMask` function.

File: vespa/test_integration_docker.py (hunk 5 above)
Comment (lesters) [reply vespa/test_integration_docker.py:212]: Documentation for that is on https://docs.vespa.ai/en/reference/rank-features.html

**Review by lesters — APPROVED** [review]
LGTM. Even verified that docker ONNX test worked locally. Consider the comments, but not required.

File: vespa/test_integration_docker.py (hunk 1 above)
Comment (thigm85) [reply vespa/test_integration_docker.py:245]: I tried to change, and I get the following message: 
```
WARNING: invalid rank feature 'onnx(bert_tiny).logits': unknown basename: 'onnx'\n
```
I thought this might be because I had an old Vespa image, but the same error is displayed when I deploy to hosted vespa.

**Review by oyving — APPROVED** [review]
LGTM - just added a minor shorter suggestion on how to deal with default `None` values

File: vespa/package.py [hunk 8]
```diff
@@ -495,6 +581,10 @@ def __init__(
                 rank_profile.name: rank_profile for rank_profile in rank_profiles
             }
 
+        self.models = []
+        if models is not None:
+            self.models = [x for x in models]
```
[the remark this answers is not in the document]
Comment (thigm85) [reply vespa/package.py:586]: Great suggestion. I assume you meant `list(models)`

File: vespa/test_integration_docker.py (hunk 5 above)
Comment (thigm85) [reply vespa/test_integration_docker.py:212]: I will change the expressions to `tokenInputIds` in a new PR once I get confirmed the [bug fix](https://github.com/vespa-engine/vespa/pull/16253) got released and worked as expected.
